File: backend/src/services/azureTextToSpeech.ts
import { injectable } from 'tsyringe';
import { ITextToSpeech } from '../interfaces/ITextToSpeech';
import { ResultReason, SpeechConfig, SpeechSynthesizer } from 'microsoft-cognitiveservices-speech-sdk';

@injectable()
class AzureTextToSpeech implements ITextToSpeech {
    private speechConfig: SpeechConfig;

    constructor() {
        this.speechConfig = SpeechConfig.fromSubscription(process.env.AZURE_SPEECH_KEY, process.env.AZURE_SPEECH_REGION);
        this.speechConfig.speechSynthesisVoiceName = "ja-JP-AoiNeural";
    }

    public async synthesize(text: string): Promise<Buffer> {

        const synthesizer = new SpeechSynthesizer(this.speechConfig);

        const result = await new Promise<Buffer>((resolve, reject) => {
            synthesizer.speakTextAsync(
                text,
                result => {
                    const { audioData } = result;

                    synthesizer.close();
                    if (result.reason === ResultReason.SynthesizingAudioCompleted) {
                        resolve(Buffer.from(audioData));
                    } else {
                        reject(new Error(`Recognition failed: ${result.errorDetails}`));
                    }
                }, error => {
                    synthesizer.close();
                    reject(new Error(`Recognition error: ${error}`));
                });
        });

        return result;
    }
}

export { AzureTextToSpeech };

File: frontend/src/azure_text_to_speech.py
import os
import azure.cognitiveservices.speech as speechsdk
from abstract.abstract_text_to_speech import AbstractTextToSpeech

class AzureTextToSpeech(AbstractTextToSpeech):
    def __init__(self) -> None:
        self.speech_key = os.getenv('AZURE_SPEECH_KEY')
        self.speech_region = os.getenv('AZURE_SPEECH_REGION')
        self.speech_voice_name = os.getenv('AZURE_SPEECH_VOICE_NAME')

    def synthesize(self, text: str) -> None:
        
        speech_config = speechsdk.SpeechConfig(subscription=self.speech_key, region=self.speech_region)
        audio_config = speechsdk.audio.AudioOutputConfig(use_default_speaker=True)
        speech_config.speech_synthesis_voice_name=self.speech_voice_name

        speech_synthesizer = speechsdk.SpeechSynthesizer(speech_config=speech_config, audio_config=audio_config)

        # 音声合成、再生
        speech_synthesis_result = speech_synthesizer.speak_text_async(text).get()

        print(speech_synthesis_result.reason)

        return

